api/envoy/admin/v3/server_info.proto

```diff
@@ -79,6 +79,14 @@ message CommandLineOptions {
     InitOnly = 2;
   }
 
+  enum DefaultsProfile {
+    // Default option: run with settings optimized for performance.
+    Performant = 0;
```

Comment:
High level first pass.  
Before we add more complexity to the API, have we loadtested with performant vs safe to see if there's a measurable difference?
If not I'd be inclined to say we should loadtest, and if there's not a visible difference, we should just (slowly, abiding by standard API policy rules) change Envoy's defaults to the safe defaults.

Reply:
I think this is a great thing to call out. Have we looked at this?

If we think there is a big perf difference, my preference would be to somehow default to secure with some way to runtime flip back to performant (I'm not sure if this is actually practical given when things load.)?

Would like to have this discussion before doing a full review.

/wait-any

Reply:
Yeah, it seems some safe defaults should "just work" as regular defaults. However, in general, we are going to have places where we need to make tradeoffs between safety, stability and performance. Examples would include in a "secure" profile the ability to turn on more aggressive path and header normalization and defaults in secure mode.

I think the more general picture for these profiles is that we could have sensible defaults made available to Envoy users who expect different environments, e.g. cloud vs. public Internet (there's almost certainly different sensible defaults given latency and BDP differences there).

Regarding API, I don't think this work really adds complexity to the API. Each xDS client is free to set its own defaults, so this is really orthogonal, it's a question of how Envoy is opting to do its defaults.

Profiles also allow us to make a more granular decision on defaults for different subsets of users. E.g. if we introduce a `paranoid_security` profile, users opting will always have the most secure known settings, even at the risk of breaking existing traffic, because workload protection is more important.

So, happy to have discussion, and I think we should default secure, but I'm still pretty keen to see profiles land.

Reply:
Yes, these exactly my thoughts. I think naming the default profile "performant" made this PR to be looked at in a very narrow sense. (This was my suggestion, as performance issues were raised back when we were fixing Netflix CVEs and there was a concern that enabling edge settings by default may drop performance).

We did not not do load testing. I do not see a load test that can give a simple answer whether the default defaults are faster than edge defaults. I think some parameters do not affect performance, like timeouts, and some may affect load tests with specific environments.

And I think performance is not the most important aspect of this feature and it should not be viewed as just separation between performance and safety. For example we have a number of Envoy based products with different roles configured by the same config pipeline. These roles require different dataplane config based on use case (i.e. envoy within service mesh with trusted traffic, vs edge proxies, size of proxies, etc). Also the dataplane config for some products may need to be changed fast in response to CVE. It is much easier to bake it into the binary and roll it out when having to change common control plane.

In some sense this PR proposes a separation from low level dataplane concerns that are best managed by Envoy developers, from higher level configuration concerns that product teams focus on.

Reply:
I like the idea of profiles so I'm in favor of landing them, thanks for the discussion.

I would prefer to have "safe" by the default profile, mainly because there is less foot gun risk and it then makes users explicitly turn off safety. We are gradually moving Envoy to safe defaults so this seems like a natural progression? I'm mainly wondering if we should think about what the new defaults are for Envoy as part of this change? Since once we introduce profiles it might be a bit harder to flip the default (maybe not, but worth discussing.)

Reply:
No objection to profiles -  I think it'll make it much easier to have correct defaults across varied deployments.  I just think most of the defaults set in this PR should simply be Envoy default. We could do that by making "safe" be Envoy's default as Matt suggests above, and only want to keep the old options around if they're actually more performant.  If they're not, we could have a test-only alternate profile which would have the advantages of not codifying old config (if there's no reason to keep it around) and having an example of how to add your own defaults not in core code.  WDYT?

Reply:
SG